### PR TITLE
Gracefully handle missing headers

### LINKED ISSUE #1781: TypedHeader shouldn't reject missing headers by default
- [x] I have looked for existing issues (including closed) about this

## Feature Request

TypedHeader should allow specific Header type to handle missing headers instead of immediately rejecting the request.

### Motivation

`Header` trait already allows to decode headers from arbitrary iterators, including iterators with zero, one or more header value repetitions.

Some `Header` implementations require their header value to be always present, in which case `Header::decode` will return an error anyway, while many other headers might be optional by design, in which case their `Header::decode` will correctly handle empty iterator and return some default parsed value (e.g. empty collection).

However, right now axum's `TypedHeader` checks for header name's presence in the request and, if it's not there, immediately rejects the request, not giving those `Header::decode` implementation a chance to gracefully handle the empty case.

### Proposal

Forward empty iterator to the `<H as Header>::decode` and use its result instead of immediately rejecting the request. This will correctly handle both mandatory headers (as those will still return an error) as well as optional ones.

### Alternatives

If this is considered undesirable for some reason, another alternative could be to implement a custom extractor for `Option<TypedHeader<H>>`. It won't be ideal as the user will have to manually invoke `H::decode` again to get the correct default value, but might be less of a breaking change.

### PR DESCRIPTION
**Pull request by RReverser**
Reports a missing header rejection only if the header's decoder didn't return a default value.

I've also added a regression test by using Cookie as an example: missing Cookie should be treated in the same way as empty Cookie header, but before this PR would reject the entire request. Now it's going to be handled correctly.

Closes #1781.

### CODE REVIEW

File: axum/src/typed_header.rs [hunk 1]
```diff
@@ -61,17 +61,19 @@ where
...
-                reason: TypedHeaderRejectionReason::Missing,
-            }),
-            Err(err) => Err(TypedHeaderRejection {
-                name: T::name(),
-                reason: TypedHeaderRejectionReason::Error(err),
-            }),
-        }
+                reason: if is_missing {
+                    // Report a more precise rejection for the missing header case.
+                    TypedHeaderRejectionReason::Missing
```
Comment (jplatte) [inline axum/src/typed_header.rs:72]: I'd be fine removing this variant, personally.

**Review by jplatte — COMMENTED** [review]
Looks good, but shouldn't this be based on the 0.7.0 branch?

**Review by jplatte — COMMENTED** [review]
If @davidpdrsn agrees, I think we can merge this as a non-breaking change.

File: axum/src/typed_header.rs (hunk 1 above)
Comment (RReverser) [reply axum/src/typed_header.rs:72]: Yeah I was a bit torn.

The downsides of doing so are:

1. Slightly worse error message / debugging experience.
2. That kind of change would certainly be breaking as users might be checking directly against the error enum.

It seems fairly low-cost to keep it for backward compat and for the slightly better error message for now.

File: axum/src/typed_header.rs [hunk 2]
```diff
@@ -174,19 +176,37 @@ mod tests {
...
+            .get("/")
+            .header("user-agent", "foobar")
+            .header("content-encoding", "gzip, br")
+            .header("content-encoding", "deflate")
+            .send()
+            .await;
+        let body = res.text().await;
+        assert_eq!(
+            body,
+            r#"UserAgent("foobar"), ContentEncoding("gzip, br, deflate")"#
```
[the remark this answers is not in the document]
Comment (RReverser) [reply axum/src/typed_header.rs:198]: That's probably true; unfortunately, in the Content-Encoding case they literally don't expose all the values in any other way :/

I could add some custom `Header` implementation here instead but not sure it's worth it.

[the remark this answers is not in the document]
Comment (jplatte) [reply axum/src/typed_header.rs:198]: Okay. Maybe it's okay for now, but sounds like `ContentEncoding` should be improved upstream ^^

[the remark this answers is not in the document]
Comment (RReverser) [reply axum/src/typed_header.rs:198]: I probably could just use `.contains` for each of the content-encodings separately, but that's not as exhaustive as checking the inner list.

[the remark this answers is not in the document]
Comment (RReverser) [reply axum/src/typed_header.rs]: Changed to Cookie. I'm still using debug representation for convenience, but no longer of the external opaque types, but of things like `&str` and `Vec<(&str, &str)>` which should be stable.

File: axum/src/typed_header.rs (hunk 1 above)
Comment (davidpdrsn) [reply axum/src/typed_header.rs:72]: I think we should keep it. For example for auth headers it feels valuable to be able to distinguish a missing header from an invalid header value. Stuff like https://github.com/tokio-rs/axum/issues/316 comes to mind.

**Review by davidpdrsn — COMMENTED** [review]
Looks good to me! Wanna update the change log as well?

**Comment by RReverser** [comment]
> Looks good to me! Wanna update the change log as well?

Done.

**Review by davidpdrsn — APPROVED** [review]
Thanks!
